src/terms/local.jl

```diff
@@ -71,19 +71,32 @@ function (external::ExternalFromFourier)(basis::PlaneWaveBasis{T}) where {T}
 end
 
 # Returns the form factors at unique values of |G + q| (in Cartesian coordinates).
-# Uses a hash map for O(1) lookup.
+# Additionally, returns a mapping from any G index to a unique |G + q|
 function atomic_local_form_factors(basis::PlaneWaveBasis{T}, Gqs_cart::AbstractArray) where{T}
-    form_factors = IdDict{Tuple{Int,T},T}()  # IdDict for Dual compatibility
-    for G in Gqs_cart
+    iG2norm = zeros(Int, length(Gqs_cart))
+    norms = IdDict{T, Int}()
+    for (iG, G) in enumerate(Gqs_cart)
         p = norm(G)
+        if !haskey(norms, p)
+            norms[p] = length(norms) + 1
+        end
+        iG2norm[iG] = norms[p]
+    end
+
+    form_factors = zeros(T, length(norms), length(basis.model.atom_groups))
+    for (iG, G) in enumerate(Gqs_cart)
+        p = norm(G)
+        inorm = iG2norm[iG]
         for (igroup, group) in enumerate(basis.model.atom_groups)
-            if !haskey(form_factors, (igroup, p))
+            if form_factors[inorm, igroup] ≈ zero(T)
```

Comment:
it should be exactly zero, no ?

It is unexepected to me to equal "zero" and "unitialised", I'm not yet sure what to think about it right now. But let's just think again if there is another way.

Reply:
I'm always reluctant to use exact comparison with floats, but maybe this is fine in Julia. As for the "zero" = "uninitialized", I'll give it some thoughts.

Reply:
I came up with a work around. It's a bit clunky, but more explicit.

Reply:
I think this goes in the right direction, I made another suggestion.